### PR TITLE
IGNITE-22069 Optimize entries expiration

### PR DESCRIPTION
**Pull request by alex-plekhanov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java [hunk 1]
```diff
@@ -1108,7 +1110,30 @@ private Metas getOrAllocateCacheMetas() throws IgniteCheckedException {
...
             int cleared = 0;
 
-            for (CacheDataStore store : cacheDataStores()) {
+            Iterable<CacheDataStore> cacheDataStores = cacheDataStores();
+
+            // Use random shift to reduce contention.
+            int shift = ThreadLocalRandom.current().nextInt(F.size(cacheDataStores.iterator()));
+
+            int cnt = 0;
+            for (CacheDataStore store : cacheDataStores) {
```
Comment (yurinaryshkin) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java:1119]: This doesn't seem to work. After executing F.size() cacheDataStores doesn't return elements.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/expiry/IgniteCacheExpiryPolicyTestSuite.java [hunk 2]
```diff
@@ -57,7 +57,8 @@
 
     IgniteCacheExpireWhileRebalanceTest.class,
 
-    ExpiryPolicyInfoLoggingTest.class
+    ExpiryPolicyInfoLoggingTest.class,
```
Comment (yurinaryshkin) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/expiry/IgniteCacheExpiryPolicyTestSuite.java:60]: Is there a test that checks ttl-cleanup-worker for persistence database? I'd propose to enhance GridCacheTtlManagerSelfTest with parameter persistence.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java [hunk 3]
```diff
@@ -2169,6 +2169,20 @@ public List<L> remove(L lower, L upper, int limit) throws IgniteCheckedException
...
 
+    /**
+     * @param lower Lower bound (inclusive).
+     * @param upper Upper bound (inclusive).
+     * @param x Implementation specific argument.
+     * @param limit Limit of processed entries by single call, {@code 0} for no limit.
+     * @return {@code True} if removed at least one row.
+     * @throws IgniteCheckedException If failed.
+     */
+    protected boolean removex(L lower, L upper, Object x, int limit) throws IgniteCheckedException {
```
Comment (yurinaryshkin) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java:2180]: Minor: is this method going to be used from several places? Should it be inlined?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java (hunk 1 above)
Comment (alex-plekhanov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java]: Thanks, fixed

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/expiry/IgniteCacheExpiryPolicyTestSuite.java (hunk 2 above)
Comment (alex-plekhanov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/expiry/IgniteCacheExpiryPolicyTestSuite.java:60]: Parameter added to the test

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java (hunk 3 above)
Comment (alex-plekhanov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java:2180]: This method uses private class and private method, so can't be inlined in other classes.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java [hunk 4]
```diff
@@ -2761,43 +2774,37 @@ private int purgeExpiredInternal(
...
-                                c.apply(e1, obsoleteVer);
+                            if (entry != null)
+                                c.apply(entry, obsoleteVer);
                         }
 
-                        cleared++;
+                        cleared += rows.size();
                     }
-                    while (cur.next());
+                    while (amount < 0 || cleared < amount);
```
Comment (yurinaryshkin) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java:2807]: What is this condition for ?
`amount < 0`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java [hunk 5]
```diff
@@ -2169,6 +2169,20 @@ public List<L> remove(L lower, L upper, int limit) throws IgniteCheckedException
         return Collections.unmodifiableList(rmvOp.removedRows);
     }
 
+    /**
+     * @param lower Lower bound (inclusive).
+     * @param upper Upper bound (inclusive).
+     * @param x Implementation specific argument.
+     * @param limit Limit of processed entries by single call, {@code 0} for no limit.
```
Comment (yurinaryshkin) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java:2176]: Both invocations of the method pass -1 as parameter limit. Should this value be added to the Javadoc?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java (hunk 4 above)
Comment (alex-plekhanov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java:2807]: See javadoc for `IgniteCacheOffheapManager#expire`, -1 as parameter assume no limit. We never use it in code, but it was supported in the previous implementation and it's easy to support now, so I decided to keep this condition.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java (hunk 5 above)
Comment (alex-plekhanov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java:2176]: Ok

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java [hunk 6]
```diff
@@ -1108,7 +1110,28 @@ private Metas getOrAllocateCacheMetas() throws IgniteCheckedException {
         try {
             int cleared = 0;
 
+            // Use random shift to reduce contention.
+            int shift = ThreadLocalRandom.current().nextInt(F.size(cacheDataStores().iterator()));
+
+            int cnt = 0;
+            for (CacheDataStore store : cacheDataStores()) {
```
Comment (yurinaryshkin) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java:1117]: As far as I understood, the logic here is to split cacheDataStores() into 2 parts at random index and after that process right sublist before the left sublist. Can this change be simplified to minimize code duplication? For example, construct a new list from 2 sublists and process it as before.

**Review by yurinaryshkin — APPROVED** [review]

**Pull request closed without merging**